File: public/python/007-merge-intervals.py
# LeetCode Problem #56: Merge Intervals
# Difficulty: Medium
# Link: https://leetcode.com/problems/merge-intervals/

# ─────────────────────────────────────────────
# APPROACH 1: Brute Force | O(n³) time | O(n) space
# EXPLAIN: Repeatedly scan all interval pairs and merge overlapping ones until no merges occur.
# WHEN: Never (too inefficient) — illustrates why sorting is necessary.

def merge_intervals_brute(intervals: list[list[int]]) -> list[list[int]]:
    result = [iv[:] for iv in intervals]
    merged = True
    while merged:
        merged = False
        temp: list[list[int]] = []
        used = [False] * len(result)
        for i in range(len(result)):
            if used[i]:
                continue
            cur = result[i][:]
            used[i] = True
            for j in range(i + 1, len(result)):
                if used[j]:
                    continue
                if cur[1] >= result[j][0] and cur[0] <= result[j][1]:
                    cur[0] = min(cur[0], result[j][0])
                    cur[1] = max(cur[1], result[j][1])
                    used[j] = True
                    merged = True
            temp.append(cur)
        result = temp
    return result


# ─────────────────────────────────────────────
# APPROACH 2: Sort + Linear Merge | O(n log n) time | O(n) space
# EXPLAIN: Sort intervals by start time, then greedily merge each interval into the last output interval if they overlap.
# WHEN: Standard and clean — the go-to approach for most interview and production scenarios.

def merge_intervals_sort(intervals: list[list[int]]) -> list[list[int]]:
    if not intervals:
        return []
    intervals.sort(key=lambda x: x[0])
    merged = [intervals[0]]
    for start, end in intervals[1:]:
        if start <= merged[-1][1]:              # overlaps with the last merged interval
            merged[-1][1] = max(merged[-1][1], end)
        else:
            merged.append([start, end])
    return merged


# ─────────────────────────────────────────────
# APPROACH 3: Sorting with Lambda | O(n log n) time | O(n) space
# EXPLAIN: Same as sort + merge but uses a running start/end pair instead of mutating the last list element.
# WHEN: Cleaner functional style; avoids in-place mutation of the result list.

def merge_intervals_functional(intervals: list[list[int]]) -> list[list[int]]:
    if not intervals:
        return []
    intervals.sort(key=lambda x: x[0])
    result: list[list[int]] = []
    start, end = intervals[0]
    for s, e in intervals[1:]:
        if s <= end:
            end = max(end, e)
        else:
            result.append([start, end])
            start, end = s, e
    result.append([start, end])
    return result


# ─────────────────────────────────────────────
# APPROACH 4: Without Modifying Result | O(n log n) time | O(n) space
# EXPLAIN: Decompose each interval into +1 (start) and -1 (end) events, sweep through counting active intervals.
# WHEN: When you need to generalise to counting overlapping intervals or detecting the maximum overlap count.

def merge_intervals_sweep(intervals: list[list[int]]) -> list[list[int]]:
    events: list[tuple[int, int]] = []
    for start, end in intervals:
        events.append((start,  1))
        events.append((end,   -1))
    # At equal times process opens (1) before closes (-1)
    events.sort(key=lambda e: (e[0], -e[1]))

    result: list[list[int]] = []
    active = 0
    seg_start = 0
    for time, delta in events:
        if active == 0 and delta == 1:
            seg_start = time
        active += delta
        if active == 0:
            result.append([seg_start, time])
    return result


# ─────────────────────────────────────────────
if __name__ == "__main__":
    cases = [
        ([[1,3],[2,6],[8,10],[15,18]], [[1,6],[8,10],[15,18]]),
        ([[1,4],[4,5]],                [[1,5]]),
        ([[1,4],[2,3]],                [[1,4]]),
    ]
    import copy
    for intervals, expected in cases:
        assert merge_intervals_brute(copy.deepcopy(intervals))      == expected
        assert merge_intervals_sort(copy.deepcopy(intervals))       == expected
        assert merge_intervals_functional(copy.deepcopy(intervals)) == expected
        assert merge_intervals_sweep(copy.deepcopy(intervals))      == expected
    print("All tests passed.")

# Made with Bob

File: src/data/tutorials/007-merge-intervals.js
/**
 * Tutorial content for #007 — Merge Intervals
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Given a collection of intervals, merge all overlapping intervals and return the result. Two intervals overlap if one starts before the other ends.`,
    example: `intervals = [[1,3],[2,6],[8,10],[15,18]]\n→ [1,3] and [2,6] overlap (3 ≥ 2) → merge to [1,6]\n→ [8,10] and [15,18] don't overlap\n✅ Answer: [[1,6],[8,10],[15,18]]`,
    keyInsight: `After sorting by start time, you only ever need to check if the current interval overlaps with the last interval in your result list — no need to check all previous ones.`,
  },

  approaches: {
    'Brute Force': {
      intuition: `Repeatedly scan the list for any two intervals that overlap and merge them into one. Keep repeating until no more merges are possible. Very simple to reason about but highly inefficient — each full scan only fixes one merge.`,
      steps: [
        `Set merged = true to enter the loop.`,
        `While merged is true: set merged = false.`,
        `  Compare every pair (i, j) of intervals.`,
        `  If they overlap (max(start_i, start_j) <= min(end_i, end_j)):`,
        `    Replace both with one merged interval, set merged = true, break inner loops.`,
        `Return the final list.`,
      ],
      example: `intervals = [[1,3],[2,6],[8,10],[15,18]]\n\nPass 1: [1,3] and [2,6] overlap → merge → [[1,6],[8,10],[15,18]], merged=true\nPass 2: check all pairs — none overlap → merged=false, stop\n✅ Answer: [[1,6],[8,10],[15,18]]`,
      keyInsight: `O(n²) to O(n³) time, O(n) space. Each pass may merge only one pair, so up to n passes of O(n²) scanning. Never use this in practice.`,
    },

    'Sorting': {
      intuition: `Sort all intervals by their start time. Then walk through them once: if the current interval overlaps with the last merged interval, extend the last one's end. Otherwise, start a new interval in the result. Sorting guarantees that any overlap can only be with the immediately preceding interval.`,
      steps: [
        `Sort intervals by start value.`,
        `Initialize result list with the first interval.`,
        `For each subsequent interval curr:`,
        `  Let last = result's last interval.`,
        `  If curr.start <= last.end: they overlap → update last.end = max(last.end, curr.end).`,
        `  Else: no overlap → add curr to result as a new interval.`,
        `Return result.`,
      ],
      example: `intervals sorted: [[1,3],[2,6],[8,10],[15,18]]\n\nresult = [[1,3]]\ncurr=[2,6]: 2<=3 → merge → result=[[1,6]]\ncurr=[8,10]: 8>6 → new → result=[[1,6],[8,10]]\ncurr=[15,18]: 15>10 → new → result=[[1,6],[8,10],[15,18]]\n✅ Answer: [[1,6],[8,10],[15,18]]`,
      keyInsight: `O(n log n) time (sort dominates), O(n) space. This is the standard solution. After sorting, one linear pass is all you need — each interval is looked at exactly once.`,
    },

    'Sorting with Lambda': {
      intuition: `Identical logic to the Sorting approach, but uses a lambda/comparator explicitly passed to the sort function. This is common in Java or when you want a custom sort order that isn't the default. Same algorithm, different syntax.`,
      steps: [
        `Sort using a lambda: Arrays.sort(intervals, (a, b) -> a[0] - b[0]).`,
        `Initialize result list with the first interval.`,
        `For each subsequent interval: compare curr[0] with the last interval's end.`,
        `If overlapping: last[1] = max(last[1], curr[1]).`,
        `Else: add curr to result.`,
        `Convert result list to array and return.`,
      ],
      example: `intervals sorted via lambda: [[1,3],[2,6],[8,10],[15,18]]\n\nresult = [[1,3]]\n[2,6]: 2<=3 → last becomes [1,6]\n[8,10]: 8>6 → add new → [[1,6],[8,10]]\n[15,18]: 15>10 → add new → [[1,6],[8,10],[15,18]]\n✅ Answer: [[1,6],[8,10],[15,18]]`,
      keyInsight: `Same O(n log n) time, O(n) space as the Sorting approach. The lambda comparator is idiomatic Java — (a,b) -> a[0]-b[0] sorts by start index in ascending order.`,
    },

    'Without Modifying Result': {
      intuition: `Sort by start time, then merge intervals by tracking a "current" interval separately rather than mutating the last element of the result list. When a new interval doesn't overlap with the current one, flush the current to the result and move on. This avoids in-place list mutation, making the logic slightly easier to reason about.`,
      steps: [
        `Sort intervals by start time.`,
        `Set currStart = intervals[0][0], currEnd = intervals[0][1].`,
        `For each subsequent interval [s, e]:`,
        `  If s <= currEnd: currEnd = max(currEnd, e) (extend the current interval).`,
        `  Else: add [currStart, currEnd] to result, then set currStart=s, currEnd=e.`,
        `After the loop, add the final [currStart, currEnd] to result.`,
        `Return result.`,
      ],
      example: `intervals sorted: [[1,3],[2,6],[8,10],[15,18]]\ncurrStart=1, currEnd=3\n\n[2,6]: 2<=3 → currEnd=max(3,6)=6\n[8,10]: 8>6 → add [1,6], currStart=8, currEnd=10\n[15,18]: 15>10 → add [8,10], currStart=15, currEnd=18\nEnd of loop → add [15,18]\n✅ Answer: [[1,6],[8,10],[15,18]]`,
      keyInsight: `O(n log n) time, O(n) space — same complexity as the standard Sorting approach. The difference is stylistic: tracking current start/end variables avoids accessing the tail of the result list, which some find cleaner.`,
    },
  },
}
